rust/lance-index/src/util/datafusion.rs

```diff
@@ -239,6 +239,41 @@ pub fn safe_coerce_scalar(value: &ScalarValue, ty: &DataType) -> Option<ScalarVa
             _ => None,
         },
         ScalarValue::Null => Some(value.clone()),
+        ScalarValue::List(vals, _) => {
+            if let DataType::FixedSizeList(_, size) = ty {
+                if let Some(vals) = vals {
+                    if vals.len() as i32 != *size {
+                        return None;
+                    }
+                }
+            }
+            let (new_values, field) = match ty {
+                DataType::List(field)
+                | DataType::LargeList(field)
+                | DataType::FixedSizeList(field, _) => {
+                    if let Some(vals) = vals {
+                        let values = vals
+                            .iter()
+                            .map(|val| safe_coerce_scalar(val, field.data_type()))
+                            .collect::<Option<Vec<_>>>();
+                        (values, field)
+                    } else {
+                        (None, field)
+                    }
+                }
+                _ => return None,
+            };
+
+            match ty {
+                DataType::List(_) => {
+                    new_values.map(|new_values| ScalarValue::List(Some(new_values), field.clone()))
```

Comment:
Would `ScalarValue::List(new_values, field.clone())` be equivalent to `new_values.map(|new_values| ScalarValue::List(Some(new_values), field.clone()))`? I think the current code maps a null scalar to "cannot coerce" (though I'm not sure if a null scalar like this is even possible at this point since I think it'd be converted to `ScalarValue::Null`)?

Reply:
This is a good point. I've updated this.